### ISSUE TITLE
Locale issues

### ISSUE THREAD

**Issue by usualstuff** [bitnami/bitnami-docker-postgresql#74]
**Description**

By default pg_dump will save CREATE DATABASE statement with locales. When executing this statement on bitnami postgresql container, it will throw an error, in my case it was: 
`ERROR:  invalid locale name: "en_US.utf8"`

Current workaround that worked for me is to replace all locales with default fallback for linux: "C".

I think something like `locale-gen` with popular locales should be enough to fix this issue.

**Steps to reproduce the issue:**

1. [First Step]
Execute this in fresh bitnami postgres docker container: 
`CREATE DATABASE example WITH TEMPLATE = template0 ENCODING = 'UTF8' LC_COLLATE = 'en_US.utf8' LC_CTYPE = 'en_US.utf8';`

**Describe the results you received:**
Postgresql response:
`ERROR:  invalid locale name: "en_US.utf8"`

**Additional information you deem important (e.g. issue happens only occasionally):**

**Version**

- Output of `docker version`:

```
 docker version
Client:
 Version:      17.05.0-ce
 API version:  1.29
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:10:54 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.05.0-ce
 API version:  1.29 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:10:54 2017
 OS/Arch:      linux/amd64
 Experimental: false

```

- Output of `docker info`:

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 6
Server Version: 17.05.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 62
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9048e5e50717ea4497b757314bad98ea3763c145
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-59-generic
Operating System: Ubuntu 16.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.61GiB
Name: redacted
ID: I3E7:3O5I:VUVD:BIZK:MTYY:4CB6:TJDO:64EU:5YBD:VOD7:73RF:GHHH
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: redacted
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support

```

- Output of `docker-compose version` (if applicable):

```
docker-compose version 1.9.0, build 2585387
docker-py version: 1.10.6
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1t  3 May 2016
```

**Comment by juan131** [comment]
Hello @usualstuff I was unable to reproduce the issue. These are the steps I follow:

 - Clone this repo.
 - Run `docker-compose up`.
 - Connect to the PostgreSQL container by running:

```
$ docker exec -it bitnamidockerpostgresql_postgresql_1 bash
```

 - Connect to PostgreSQL:

```
$ psql -U postgres
```

 - Run the command you shared:

```
postgres=# CREATE DATABASE example WITH TEMPLATE = template0 ENCODING = 'UTF8' LC_COLLATE = 'en_US.utf8' LC_CTYPE = 'en_US.utf8'
postgres=# \l
                             List of databases
   Name    |  Owner   | Encoding | Collate | Ctype |   Access privileges
-----------+----------+----------+---------+-------+-----------------------
 postgres  | postgres | UTF8     | C       | C     |
 template0 | postgres | UTF8     | C       | C     | =c/postgres          +
           |          |          |         |       | postgres=CTc/postgres
 template1 | postgres | UTF8     | C       | C     | =c/postgres          +
           |          |          |         |       | postgres=CTc/postgres
(3 rows)
```

As you can see the database is created without any issue.

**Comment by usualstuff** [comment]
@juan131 
What OS distribution you are using? I've tried to follow your steps to double check the issue but was able to reproduce it.
```
git clone https://github.com/bitnami/bitnami-docker-postgresql
cd bitnami-docker-postgresql
docker-compose up -d
docker-compose exec postgresql psql -U postgres
CREATE DATABASE example WITH TEMPLATE = template0 ENCODING = 'UTF8' LC_COLLATE = 'en_US.utf8' LC_CTYPE = 'en_US.utf8';
```
Got same error:
```
ERROR:  invalid locale name: "en_US.utf8"
```

I'm using Ubuntu 16.04.

**Comment by juan131** [comment]
Hello, I could workaround by following these steps:

```
$ git clone https://github.com/bitnami/bitnami-docker-postgresql
$ cd bitnami-docker-postgresql
$ docker-compose up -d
$ docker-compose exec postgresql dpkg-reconfigure locales
```

All the options are displayed. In my case the option `en_US.UTF-8` was the 146. When prompted about which locale to generate, I chose it:

```
Locales to be generated: 146

Many packages in Debian use locales to display text in the correct language for the
user. You can choose a default locale for the system from the generated locales.

This will select the default language for the entire system. If this system is a
multi-user system where not all users are able to speak the default language, they will
experience difficulties.

  1. None  2. C.UTF-8  3. en_US.UTF-8
Default locale for the system environment: 3

Generating locales (this might take a while)...
  en_US.UTF-8... done
Generation complete.
```

You can check that it's been generated:

```
$ docker-compose exec postgresql locale -a
C
C.UTF-8
POSIX
en_US.utf8
```

Finally:

```
docker-compose exec postgresql psql -U postgres
CREATE DATABASE example WITH TEMPLATE = template0 ENCODING = 'UTF8' LC_COLLATE = 'en_US.utf8' LC_CTYPE = 'en_US.utf8';
```

**Comment by usualstuff** [comment]
Yes, I've already used a workaround by choosing "C" locale but it would be nice if this issue would be fixed and no workaround needed.

**Comment by jotamartos** [comment]
Hi @usualstuff,

I'm glad to hear that it worked for you. We're going to open a task on our side to update the container to not to run into this issue in the future.

Regards,
Jota

**Comment by katmai** [comment]
hi, any word on this one? i'd also like to be able to change the locale and ctype

**Comment by juan131** [comment]
Hi @katmai 

Did you try the workaround I suggested to reconfigure the locale settings?

**Comment by juan131** [comment]
Hi @jozko 

As you mentioned including all the locales is not ideal since the image size will increase significantly. As you mentioned, you can create your own image based on the Bitnami one or you can also create a init script that install them once the container is created. 

For instance, you can include your code in the line 9 of the [run.sh](https://github.com/bitnami/bitnami-docker-postgresql/blob/master/10/debian-9/rootfs/run.sh#L9) file. This way the image size will remain the same and the locales will be generated during the initialisation. Please note this method will make the initialisation slower.

On Kubernetes, you could use a `initContainer` to do so too.

**Comment by juan131** [comment]
> I don't think adding commands to run.sh or dropping files into /docker-entrypoint-initdb.d/ will help - container is running with unprivileged UID 1001 at the point those are run so locale-gen won't be able to add additional locales - root permissions are required to do that.

You're right! We moved the container to a non_root approach and, therefore, those actions can't be done during the initialisation.

@beltran-rubo what do you about including `en_US.utf8` locale? I think it's worth doing to improve the usability of the image.

**Comment by flah00** [comment]
What do folks think about an init pod, which installs locales defined in an env var?

**Comment by juan131** [comment]
I guess you mean to include a initContainer on the pod which runs PostgreSQL. That solution could be valid on K8s but it doesn't fix the issue on other platforms.

**Comment by desaintmartin** [comment]
Actually, I doubt any locale installed by an initContainer be kept in the "real" container.

Anyway, on my side, I've been progressively migrating all databases to the C locale through pg_Dump / pg_restore.  Not ideal but works.

**Comment by juan131** [comment]
Hi everyone,

We just released new images of PostgreSQL (**9.6.11-r33**, **10.6.0-r34**, **11.1.0-r29**) including the locales `en_GB.UTF-8` and `en_US.UTF-8` by default.

Please feel free to test these new images and close the issue if you consider it solved with these latest images.

**Comment by alebabai** [comment]
Is it possible to configure LC_COLLATE and LC_CTYPE for default DB (that are created with name $POSTGRESQL_DATABASE)?

**Comment by alebabai** [comment]
Here is some black magic
```
RUN sed -i "s|LC_ALL: 'C'|LC_ALL: 'en_US.UTF-8'|g" /.nami/components/com.bitnami.postgresql/helpers.js
```

**Comment by juan131** [comment]
Hi @alebabai 

I'd rather make it optional. I mean, I don't want to force using a specific locale int he default database created when setting the env. variable **POSTGRESQL_DATABASE**.

We could add another env. variable to specify the locale used on that database. What do you think about that?

**Comment by alebabai** [comment]
@juan131 i just provided the workaround for the current implementation.
Better way to solve this issue is to provide env variable with initdb flags like in original postgres docker image. I've already created issue for that #107

**Comment by juan131** [comment]
Great @alebabai!! Let's continue the discussion for this workaround in the other issue you opened.
